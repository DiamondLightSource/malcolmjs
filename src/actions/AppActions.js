/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Dispatcher from '../core/Dispatcher';
import ActionTypes from '../constants/ActionTypes';
import ExecutionEnvironment from 'react/lib/ExecutionEnvironment';
import http from 'superagent';

export default {

  navigateTo(path, options) {
    if (ExecutionEnvironment.canUseDOM) {
      if (options && options.replace) {
        window.history.replaceState({}, document.title, path);
      } else {
        window.history.pushState({}, document.title, path);
      }
    }

    Dispatcher.handleViewAction({
      actionType: ActionTypes.CHANGE_LOCATION,
      path
    });
  },

  loadPage(path, cb) {
    Dispatcher.handleViewAction({
      actionType: ActionTypes.LOAD_PAGE,
      path
    });

    http.get('/api/page' + path)
      .accept('application/json')
      .end((err, res) => {
        Dispatcher.handleServerAction({
          actionType: ActionTypes.LOAD_PAGE,
          path,
          err,
          page: res.body
        });
        if (cb) {
          cb();
        }
      });
  }

};
