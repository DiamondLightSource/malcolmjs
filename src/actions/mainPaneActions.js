/**
 * Created by twi18192 on 25/08/15.
 */

import AppDispatcher from '../dispatcher/appDispatcher.js';
import appConstants  from '../constants/appConstants.js';

class MainPaneActions {
  constructor()
    {

    }

  toggleFooter1(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.FOOTER_TOGGLE,
      item: item
    })
  }

  toggleConfigPanel(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.CONFIG_TOGGLE,
      item: item
    })
  }
  toggleFavPanel(item){
    AppDispatcher.handleViewAction({
      actionType: appConstants.FAV_TOGGLE,
      item: item
    })
  }
}

const mainPaneActions = new MainPaneActions();

export default mainPaneActions;
