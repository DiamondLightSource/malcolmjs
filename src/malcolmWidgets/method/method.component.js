import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';
import Typography from '@material-ui/core/Typography';
import Tooltip from '@material-ui/core/Tooltip';
import IconButton from '@material-ui/core/IconButton';
import AttributeAlarm, {
  AlarmStates,
} from '../attributeDetails/attributeAlarm/attributeAlarm.component';
import ButtonAction from '../buttonAction/buttonAction.component';

import GroupExpander from '../groupExpander/groupExpander.component';
import blockUtils from '../../malcolm/blockUtils';
import {
  malcolmSetFlag,
  malcolmPostAction,
} from '../../malcolm/malcolmActionCreators';
import {
  malcolmUpdateMethodInput,
  malcolmArchivePost,
} from '../../malcolm/actions/method.actions';

import {
  selectorFunction,
  getDefaultFromType,
} from '../attributeDetails/attributeSelector/attributeSelector.component';
import navigationActions from '../../malcolm/actions/navigation.actions';

const styles = () => ({
  div: {
    display: 'flex',
    paddingLeft: 5,
    paddingRight: 5,
    alignItems: 'center',
  },
  textName: {
    flexGrow: 1,
    textAlign: 'Left',
    marginLeft: 4,
    marginRight: 4,
  },
  controlContainer: {
    width: 150,
    padding: 2,
  },
  runButton: {
    flexGrow: 1,
    padding: 2,
  },
  missingAttribute: {
    color: 'red',
  },
  button: {
    width: '22px',
    height: '22px',
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
});

const buildIOComponent = (input, props, isOutput) => {
  const parameterMeta = input[1];
  const widgetTag = parameterMeta.tags.find(t => t.indexOf('widget:') !== -1);
  const flags = {
    isDisabled: props.methodPending || !input[1].writeable,
    isErrorState: props.methodErrored && input[1].writeable,
    isDirty: props.dirtyInputs[input[0]],
  };
  const updateStoreOnEveryValueChange = true;

  const valueMap = isOutput ? props.outputValues : props.inputValues;
  let inputValue;
  if (valueMap[input[0]] !== undefined) {
    inputValue = valueMap[input[0]];
  } else if (props.defaultValues[input[0]] !== undefined) {
    inputValue = props.defaultValues[input[0]];
  } else {
    inputValue = getDefaultFromType(input[1]);
    if (inputValue !== undefined) {
      props.updateInput(props.methodPath, input[0], inputValue);
    }
  }
  const submitHandler = (path, value) => {
    props.updateInput(path, input[0], value);
  };
  const setFlag = (path, flagName, isDirty) => {
    props.updateInput(path, input[0], { isDirty });
  };
  const subElement = isOutput ? `returns.${input[0]}` : `takes.${input[0]}`;
  if (widgetTag) {
    return selectorFunction(
      widgetTag,
      [...props.methodPath, subElement],
      inputValue,
      submitHandler,
      flags,
      setFlag,
      props.theme.palette.primary.light,
      parameterMeta,
      false,
      updateStoreOnEveryValueChange,
      props.methodParamClickHandler
    );
  }
  return selectorFunction('widget:undefined');
};

const MethodDetails = props => {
  if (
    (!props.inputs || !Object.keys(props.inputs).length) &&
    (!props.outputs || !Object.keys(props.outputs).length)
  ) {
    return (
      <div className={props.classes.div}>
        <Tooltip title={props.methodErrorMessage}>
          <IconButton
            className={props.classes.button}
            disableRipple
            onClick={() =>
              props.buttonClickHandler(props.blockName, props.attributeName)
            }
          >
            <AttributeAlarm alarmSeverity={props.methodAlarm} />
          </IconButton>
        </Tooltip>
        <div className={props.classes.controlContainer}>
          <ButtonAction
            method
            text={props.methodName}
            disabled={
              !props.writeable || props.methodAlarm === AlarmStates.PENDING
            }
            clickAction={() =>
              props.runMethod(props.methodPath, props.inputValues)
            }
          />
        </div>
      </div>
    );
  }
  return (
    <GroupExpander key={props.methodPath} groupName={props.methodName} expanded>
      <div>
        {Object.entries(props.inputs).map(input => (
          <div key={input[0]} className={props.classes.div}>
            <Tooltip title={input[1].description}>
              <IconButton
                className={props.classes.button}
                disableRipple
                onClick={() =>
                  props.buttonClickHandler(
                    props.blockName,
                    props.attributeName,
                    `takes.${input[0]}`
                  )
                }
              >
                <AttributeAlarm alarmSeverity={AlarmStates.NO_ALARM} />
              </IconButton>
            </Tooltip>
            <Typography className={props.classes.textName}>
              {input[1].label}
            </Typography>
            <div className={props.classes.controlContainer}>
              {buildIOComponent(input, props, false)}
            </div>
          </div>
        ))}
        <div className={props.classes.div}>
          <Tooltip title={props.methodErrorMessage}>
            <IconButton
              className={props.classes.button}
              disableRipple
              onClick={() =>
                props.buttonClickHandler(props.blockName, props.attributeName)
              }
            >
              <AttributeAlarm alarmSeverity={props.methodAlarm} />
            </IconButton>
          </Tooltip>
          <div className={props.classes.controlContainer}>
            <ButtonAction
              method
              text={props.methodName}
              disabled={
                !props.writeable || props.methodAlarm === AlarmStates.PENDING
              }
              clickAction={() =>
                props.runMethod(props.methodPath, props.inputValues)
              }
            />
          </div>
        </div>
        {Object.keys(props.outputValues).length !== 0 ? (
          <div>
            <Typography className={props.classes.textName}>
              Last Return:
            </Typography>
            <Divider />
            {Object.entries(props.outputs).map(output => (
              <div key={output[0]} className={props.classes.div}>
                <Tooltip title={output[1].description}>
                  <IconButton
                    className={props.classes.button}
                    disableRipple
                    onClick={() =>
                      props.buttonClickHandler(
                        props.blockName,
                        props.attributeName,
                        `returns.${output[0]}`
                      )
                    }
                  >
                    <AttributeAlarm alarmSeverity={AlarmStates.NO_ALARM} />
                  </IconButton>
                </Tooltip>
                <Typography className={props.classes.textName}>
                  {output[1].label}
                </Typography>
                <div className={props.classes.controlContainer}>
                  {buildIOComponent(output, props, true)}
                </div>
              </div>
            ))}
          </div>
        ) : null}
      </div>
    </GroupExpander>
  );
};

MethodDetails.propTypes = {
  blockName: PropTypes.string.isRequired,
  attributeName: PropTypes.string.isRequired,
  methodName: PropTypes.string.isRequired,
  methodAlarm: PropTypes.number.isRequired,
  methodErrorMessage: PropTypes.string.isRequired,
  methodPath: PropTypes.arrayOf(PropTypes.string).isRequired,
  inputs: PropTypes.shape({}).isRequired,
  inputValues: PropTypes.shape({}).isRequired,
  outputs: PropTypes.shape({}).isRequired,
  outputValues: PropTypes.shape({}).isRequired,
  runMethod: PropTypes.func.isRequired,
  classes: PropTypes.shape({
    div: PropTypes.string,
    textName: PropTypes.string,
    runButton: PropTypes.string,
    controlContainer: PropTypes.string,
    button: PropTypes.string,
  }).isRequired,
  buttonClickHandler: PropTypes.func.isRequired,
  writeable: PropTypes.bool.isRequired,
};

const EMPTY = '';
const EMPTY_ARRAY = [];
const EMPTY_OBJECT = {};
const mapStateToProps = (state, ownProps) => {
  let method;
  if (ownProps.attributeName && ownProps.blockName) {
    method = blockUtils.findAttribute(
      state.malcolm.blocks,
      ownProps.blockName,
      ownProps.attributeName
    );
  }

  let alarm = AlarmStates.NO_ALARM;
  alarm =
    method && method.calculated.errorState ? AlarmStates.MAJOR_ALARM : alarm;
  alarm = method && method.calculated.pending ? AlarmStates.PENDING : alarm;

  return {
    writeable: method ? method.raw.writeable : false,
    methodName: method ? method.raw.label : 'Not found',
    methodAlarm: alarm,
    methodPending: method ? method.calculated.pending : false,
    methodErrored: method ? method.calculated.errorState : false,
    methodErrorMessage:
      method && method.calculated.errorMessage
        ? method.calculated.errorMessage
        : EMPTY,
    methodPath: method ? method.calculated.path : EMPTY_ARRAY,
    inputs: method ? method.raw.takes.elements : EMPTY_OBJECT,
    inputValues: (method && method.calculated.inputs) || EMPTY_OBJECT,
    dirtyInputs: (method && method.calculated.dirtyInputs) || EMPTY_OBJECT,
    outputs: method ? method.raw.returns.elements : EMPTY_OBJECT,
    outputValues: (method && method.calculated.outputs) || EMPTY_OBJECT,
    required: method ? method.raw.required : EMPTY_OBJECT,
    defaultValues: method ? method.raw.defaults : EMPTY_OBJECT,
  };
};

export const mapDispatchToProps = dispatch => ({
  buttonClickHandler: (blockName, attributeName, subElement) => {
    dispatch(
      navigationActions.navigateToInfo(blockName, attributeName, subElement)
    );
  },
  runMethod: (path, inputs) => {
    dispatch(malcolmSetFlag(path, 'pending', true));
    dispatch(malcolmArchivePost(path, inputs));
    dispatch(malcolmPostAction(path, inputs));
  },
  updateInput: (path, inputName, inputValue) => {
    dispatch(malcolmUpdateMethodInput(path, inputName, inputValue));
  },
  methodParamClickHandler: path => {
    dispatch(navigationActions.navigateToSubElement(path[0], path[1], path[2]));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(
  withStyles(styles, { withTheme: true })(MethodDetails)
);
