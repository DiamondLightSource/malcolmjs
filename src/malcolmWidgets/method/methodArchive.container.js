import * as React from 'react';
import PropTypes from 'prop-types';
import CircularBuffer from 'circular-buffer';
import Typography from '@material-ui/core/Typography';
import Loadable from 'react-loadable';
import WidgetTable from '../table/table.component';
// import MethodPlot from './methodPlot.component';
import TabbedPanel from '../../attributeView/tabbedMiddlePanel.component';

const noOp = () => {};

const LoadableMethodPlot = Loadable({
  loader: () => import('./methodPlot.component'),
  loading: () => <Typography>Loading...</Typography>,
});

const MethodArchive = props => {
  const timeStamps = props.methodArchive.timeStamp.toarray();
  const values = props.methodArchive.value.toarray();
  const dummyAttribute = {
    raw: {
      value: {
        timeStamp: timeStamps.map(stamp => stamp.localRunTime.toISOString()),
        alarm: props.methodArchive.alarmState.toarray(),
        value: values.map(value => value.runParameters[props.selectedParam[1]]),
      },
      meta: {
        elements: {
          alarm: {
            tags: ['info:alarm'],
            label: 'Method Alarm state',
          },
          timeStamp: {
            tags: ['widget:textupdate'],
            label:
              props.selectedParam[0] === 'takes'
                ? 'Time sent'
                : 'Time received',
          },
          value: {
            tags: ['widget:textupdate'],
            label: 'Parameter value',
          },
        },
      },
    },
    calculated: {},
  };
  const dummyArchive = {
    parent: props.methodArchive.parent,
    name: props.methodArchive.name,
    timeStamp: timeStamps.map(
      stamp =>
        props.selectedParam[0] === 'takes'
          ? stamp.localRunTime
          : stamp.localReturnTime
    ),
    alarmState: props.methodArchive.alarmState.toarray(),
    value: values.map(value => value.runParameters[props.selectedParam[1]]),
  };
  return (
    <TabbedPanel
      alwaysUpdate
      openPanels={props.openPanels}
      tabLabels={['Table', 'Plot']}
      defaultTab={props.defaultTab}
    >
      <WidgetTable
        attribute={dummyAttribute}
        hideInfo
        eventHandler={noOp}
        setFlag={noOp}
        addRow={noOp}
        infoClickHandler={noOp}
        rowClickHandler={noOp}
      />
      <LoadableMethodPlot
        attribute={dummyArchive}
        openPanels={props.openPanels}
      />
    </TabbedPanel>
  );
};

MethodArchive.propTypes = {
  selectedParam: PropTypes.arrayOf(PropTypes.string).isRequired,
  methodArchive: PropTypes.shape({
    parent: PropTypes.string,
    name: PropTypes.string,
    timeStamp: PropTypes.instanceOf(CircularBuffer),
    value: PropTypes.instanceOf(CircularBuffer),
    alarmState: PropTypes.instanceOf(CircularBuffer),
  }).isRequired,
  openPanels: PropTypes.shape({
    parent: PropTypes.bool,
    child: PropTypes.bool,
  }).isRequired,
  defaultTab: PropTypes.number,
};
MethodArchive.defaultProps = {
  defaultTab: 0,
};

export default MethodArchive;
