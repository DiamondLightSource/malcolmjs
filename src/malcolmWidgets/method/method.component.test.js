import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import MethodComponent, { mapDispatchToProps } from './method.component';
import {
  malcolmPostAction,
  malcolmSetFlag,
} from '../../malcolm/malcolmActionCreators';
import { malcolmUpdateMethodInput } from '../../malcolm/actions/method.actions';

jest.mock('../../malcolm/malcolmActionCreators');
jest.mock('../../malcolm/actions/method.actions');

describe('Method component', () => {
  let shallow;
  let mockStore;
  let mount;
  let testInputs;

  const state = { malcolm: { blocks: { Test: { attributes: [] } } } };

  const defaultInputs = {
    first: 1,
    second: 'two',
  };

  const testOutputs = {
    first: {
      label: 'First output',
      tags: ['widget:led'],
    },
    second: {
      label: 'Second output',
      tags: ['widget:textupdate'],
    },
  };

  const testInputValues = {
    first: 3,
    second: 'four',
  };

  const testOutputValues = {
    first: true,
    second: 'done',
  };

  const testMethod = (
    inputs,
    outputs,
    defaults,
    inputValues,
    outputValues,
    errorMsg
  ) => {
    const method = {
      name: 'Method',
      label: 'Test',
      errorState: !!errorMsg,
      errorMessage: errorMsg,
      path: ['Test', 'Method'],
      defaults,
      inputs: inputValues,
      takes: { elements: { ...inputs } },
      returns: { elements: { ...outputs } },
      outputs: outputValues,
    };

    state.malcolm.blocks.Test.attributes = [method];

    return (
      <MethodComponent
        store={mockStore(state)}
        blockName="Test"
        attributeName="Method"
      />
    );
  };

  beforeEach(() => {
    malcolmSetFlag.mockClear();
    malcolmUpdateMethodInput.mockClear();
    malcolmPostAction.mockClear();

    mockStore = myState => ({
      getState: () => myState,
      subscribe: () => {},
      dispatch: () => {},
    });

    shallow = createShallow({ dive: true });
    mount = createMount();

    testInputs = {
      first: {
        label: 'First input',
        description: 'a test input',
        tags: ['widget:textinput'],
      },
      second: {
        label: 'Second input',
        description: 'a test input',
        tags: ['widget:textinput'],
      },
    };
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly with empty inputs', () => {
    const wrapper = shallow(testMethod(testInputs, {}, {}, {}, {}, ''));
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with default inputs', () => {
    const wrapper = shallow(
      testMethod(testInputs, {}, defaultInputs, {}, {}, '')
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with existing input values', () => {
    const wrapper = shallow(
      testMethod(testInputs, {}, {}, testInputValues, {}, '')
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with return value', () => {
    const wrapper = shallow(
      testMethod(testInputs, testOutputs, {}, {}, testOutputValues, '')
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('mapDispatchToProps updateInput signals a method input needs updating', () => {
    const dispatch = () => {};
    const dispatchProps = mapDispatchToProps(dispatch);

    dispatchProps.updateInput(['path'], 'name', 'value');
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(1);
  });

  it('calls runMethod correctly on button click', () => {
    const wrapper = mount(
      testMethod(testInputs, testOutputs, {}, testInputValues, {}, '')
    );
    wrapper
      .find('button')
      .first()
      .simulate('click');
    expect(malcolmSetFlag).toHaveBeenCalledTimes(1);
    expect(malcolmSetFlag).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'pending',
      true
    );
    expect(malcolmPostAction).toHaveBeenCalledTimes(1);
    expect(malcolmPostAction).toHaveBeenCalledWith(
      ['Test', 'Method'],
      testInputValues
    );
  });

  it('calls updateInput correctly on changed input', () => {
    const wrapper = mount(
      testMethod(testInputs, testOutputs, {}, testInputValues, {}, '')
    );
    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { value: 'test' } });
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(2);
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'first',
      'test'
    );
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'first',
      { isDirty: true }
    );
  });

  it('buildIOcomponent returns null if no widget tag', () => {
    testInputs.second.tags = [];
    const wrapper = shallow(testMethod(testInputs, {}, {}, {}, {}, ''));
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
