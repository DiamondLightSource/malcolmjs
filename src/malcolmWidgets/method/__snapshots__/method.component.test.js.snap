// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method component buildIOcomponent returns null if no widget tag 1`] = `
<WithStyles(WidgetGroupExpander)
  expanded={true}
  groupName="Test"
  key="Test,Method"
>
  <div>
    <div
      className="MethodDetails-div-1"
      key="first"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          First input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="-"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <div
      className="MethodDetails-div-1"
      key="second"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          Second input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <pure(BugReport)
          nativeColor="red"
        />
      </div>
    </div>
    <WithStyles(Paper)
      elevation={4}
      style={
        Object {
          "backgroundColor": "#3B3B3B",
          "marginBottom": "2px",
          "marginTop": "2px",
          "paddingBottom": "2px",
          "paddingTop": "2px",
        }
      }
    >
      <div
        className="MethodDetails-div-1"
      >
        <WithStyles(Tooltip)
          title=""
        >
          <div>
            <WithTheme(AttributeAlarm)
              alarmSeverity={0}
            />
          </div>
        </WithStyles(Tooltip)>
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        />
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(ButtonAction)
            clickAction={[Function]}
            disabled={false}
            text="Test"
          />
        </div>
      </div>
    </WithStyles(Paper)>
  </div>
</WithStyles(WidgetGroupExpander)>
`;

exports[`Method component renders correctly with default inputs 1`] = `
<WithStyles(WidgetGroupExpander)
  expanded={true}
  groupName="Test"
  key="Test,Method"
>
  <div>
    <div
      className="MethodDetails-div-1"
      key="first"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          First input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="1"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <div
      className="MethodDetails-div-1"
      key="second"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          Second input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="two"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <WithStyles(Paper)
      elevation={4}
      style={
        Object {
          "backgroundColor": "#3B3B3B",
          "marginBottom": "2px",
          "marginTop": "2px",
          "paddingBottom": "2px",
          "paddingTop": "2px",
        }
      }
    >
      <div
        className="MethodDetails-div-1"
      >
        <WithStyles(Tooltip)
          title=""
        >
          <div>
            <WithTheme(AttributeAlarm)
              alarmSeverity={0}
            />
          </div>
        </WithStyles(Tooltip)>
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        />
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(ButtonAction)
            clickAction={[Function]}
            disabled={false}
            text="Test"
          />
        </div>
      </div>
    </WithStyles(Paper)>
  </div>
</WithStyles(WidgetGroupExpander)>
`;

exports[`Method component renders correctly with empty inputs 1`] = `
<WithStyles(WidgetGroupExpander)
  expanded={true}
  groupName="Test"
  key="Test,Method"
>
  <div>
    <div
      className="MethodDetails-div-1"
      key="first"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          First input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="-"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <div
      className="MethodDetails-div-1"
      key="second"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          Second input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="-"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <WithStyles(Paper)
      elevation={4}
      style={
        Object {
          "backgroundColor": "#3B3B3B",
          "marginBottom": "2px",
          "marginTop": "2px",
          "paddingBottom": "2px",
          "paddingTop": "2px",
        }
      }
    >
      <div
        className="MethodDetails-div-1"
      >
        <WithStyles(Tooltip)
          title=""
        >
          <div>
            <WithTheme(AttributeAlarm)
              alarmSeverity={0}
            />
          </div>
        </WithStyles(Tooltip)>
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        />
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(ButtonAction)
            clickAction={[Function]}
            disabled={false}
            text="Test"
          />
        </div>
      </div>
    </WithStyles(Paper)>
  </div>
</WithStyles(WidgetGroupExpander)>
`;

exports[`Method component renders correctly with existing input values 1`] = `
<WithStyles(WidgetGroupExpander)
  expanded={true}
  groupName="Test"
  key="Test,Method"
>
  <div>
    <div
      className="MethodDetails-div-1"
      key="first"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          First input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="3"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <div
      className="MethodDetails-div-1"
      key="second"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          Second input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="four"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <WithStyles(Paper)
      elevation={4}
      style={
        Object {
          "backgroundColor": "#3B3B3B",
          "marginBottom": "2px",
          "marginTop": "2px",
          "paddingBottom": "2px",
          "paddingTop": "2px",
        }
      }
    >
      <div
        className="MethodDetails-div-1"
      >
        <WithStyles(Tooltip)
          title=""
        >
          <div>
            <WithTheme(AttributeAlarm)
              alarmSeverity={0}
            />
          </div>
        </WithStyles(Tooltip)>
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        />
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(ButtonAction)
            clickAction={[Function]}
            disabled={false}
            text="Test"
          />
        </div>
      </div>
    </WithStyles(Paper)>
  </div>
</WithStyles(WidgetGroupExpander)>
`;

exports[`Method component renders correctly with return value 1`] = `
<WithStyles(WidgetGroupExpander)
  expanded={true}
  groupName="Test"
  key="Test,Method"
>
  <div>
    <div
      className="MethodDetails-div-1"
      key="first"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          First input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="-"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <div
      className="MethodDetails-div-1"
      key="second"
    >
      <div>
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
      </div>
      <WithStyles(Tooltip)
        title="a test input"
      >
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          Second input
          :
           
        </WithStyles(Typography)>
      </WithStyles(Tooltip)>
      <div
        className="MethodDetails-controlContainer-3"
      >
        <WithStyles(WidgetTextInput)
          Error={false}
          Pending={true}
          Value="-"
          blurHandler={[Function]}
          continuousSend={true}
          focusHandler={[Function]}
          forceUpdate={false}
          setFlag={[Function]}
          submitEventHandler={[Function]}
        />
      </div>
    </div>
    <WithStyles(Paper)
      elevation={4}
      style={
        Object {
          "backgroundColor": "#3B3B3B",
          "marginBottom": "2px",
          "marginTop": "2px",
          "paddingBottom": "2px",
          "paddingTop": "2px",
        }
      }
    >
      <div
        className="MethodDetails-div-1"
      >
        <WithStyles(Tooltip)
          title=""
        >
          <div>
            <WithTheme(AttributeAlarm)
              alarmSeverity={0}
            />
          </div>
        </WithStyles(Tooltip)>
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        />
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(ButtonAction)
            clickAction={[Function]}
            disabled={false}
            text="Test"
          />
        </div>
      </div>
    </WithStyles(Paper)>
    <div>
      <WithStyles(Typography)
        className="MethodDetails-textName-2"
      >
        Last Return:
      </WithStyles(Typography)>
      <WithStyles(Divider) />
      <div
        className="MethodDetails-div-1"
        key="first"
      >
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          First output
          :
           
        </WithStyles(Typography)>
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(WidgetLED)
            LEDState={true}
            colorBorder="#7986cb"
            colorCenter="#7986cb"
          />
        </div>
      </div>
      <div
        className="MethodDetails-div-1"
        key="second"
      >
        <WithTheme(AttributeAlarm)
          alarmSeverity={0}
        />
        <WithStyles(Typography)
          className="MethodDetails-textName-2"
        >
          Second output
          :
           
        </WithStyles(Typography)>
        <div
          className="MethodDetails-controlContainer-3"
        >
          <WithStyles(WidgetTextUpdate)
            Text="done"
          />
        </div>
      </div>
    </div>
  </div>
</WithStyles(WidgetGroupExpander)>
`;
